Add validation failure guide for common PR labels (#357277)
microsoft/winget-pkgs: doc/ValidationFailureGuide.md

# WinGet Package Validation: Common Failure Labels and How to Fix Them

When you submit a pull request to [microsoft/winget-pkgs](https://github.com/microsoft/winget-pkgs), automated validation pipelines run a series of checks on your manifest and installers. If something fails, labels are applied to your PR to indicate what went wrong. This document explains each label, what causes it, and how to resolve it.

> [!TIP]
> Many issues can be caught before submitting by running:
> ```powershell
> winget validate <path-to-manifest>
> winget install --manifest <path-to-manifest>
> ```

> > For isolated testing, use the [SandboxTest.ps1](https://github.com/microsoft/winget-pkgs/blob/master/doc/tools/SandboxTest.md) script in Windows Sandbox.

Reviewer: Nit: The script runs on the host and opens a sandbox. Semantically, the script isn't running in the sandbox
```suggestion
> For isolated testing, use the [SandboxTest.ps1](https://github.com/microsoft/winget-pkgs/blob/master/doc/tools/SandboxTest.md) script or Windows Sandbox.
```

---

## Table of Contents

- [Status Labels](#status-labels)
- [Error Labels](#error-labels)
  - [Manifest & Path Errors](#manifest--path-errors)
  - [Installer & Binary Errors](#installer--binary-errors)
  - [URL & Domain Errors](#url--domain-errors)
  - [Installation Testing Errors](#installation-testing-errors)
  - [Dependency Errors](#dependency-errors)
  - [Other Validation Errors](#other-validation-errors)
- [Content Policy Labels](#content-policy-labels)
- [Internal Error Labels](#internal-error-labels)
- [Moderator-Applied Labels](#moderator-applied-labels)
- [Quick Reference Table](#quick-reference-table)

---

## Status Labels

These labels track the progress of your PR through the validation pipeline.

| Label | Meaning |
|---|---|
| **Azure-Pipeline-Passed** | Your manifest passed automated testing and is awaiting moderator approval. |

> | **Validation-Completed** | All checks passed. Your PR will be merged automatically. |

Reviewer: Nit: Just because the validation passes doesn't mean that it won't fail a manual review. Better edge case coverage
```suggestion
| **Validation-Completed** | All checks passed. Your PR may be merged automatically after moderator review. |
```

| **Needs-Author-Feedback** | Something needs your attention. If not addressed within 10 days, the PR will be auto-closed. |
| **Needs-Attention** | The PR has been escalated to the WinGet engineering team for investigation. |
| **Blocking-Issue** | The PR cannot be approved until the blocking issue (indicated by an accompanying error label) is resolved. |
| **Needs-CLA** | You have not signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs). The PR cannot be merged until the CLA is signed. |

---

## Error Labels

> ### Manifest & Path Errors

Reviewer: `Manifest-Version-Deprecated` seems to be missing

#### `Manifest-Validation-Error`

**What it means:** Your manifest has a syntax error or does not conform to the [manifest schema specification](https://github.com/microsoft/winget-pkgs/tree/master/doc/manifest).

**Common causes:**
- Invalid YAML syntax (bad indentation, missing colons, incorrect types)
- Missing required fields (`PackageIdentifier`, `PackageVersion`, `PackageName`, `Publisher`, etc.)
- Using a singleton manifest (deprecated in the community repo — use multi-file manifests)

**How to fix:**
```powershell
winget validate <path-to-manifest>
```

> Address all errors reported by the validator and resubmit.

Reviewer: ```suggestion
Address all reported errors and resubmit.
```
Nit: Clarity

├── <PackageIdentifier>.installer.yaml
├── <PackageIdentifier>.locale.<language>.yaml
└── <PackageIdentifier>.yaml
```

**Common causes:**
- Incorrect casing — `PackageIdentifier` and directory names are **case-sensitive**
- Filenames don't match the `PackageIdentifier`
- Manifest placed outside the `manifests/` directory

**How to fix:** Ensure the directory path and filenames exactly match your `PackageIdentifier` and `PackageVersion`.

---

#### `PullRequest-Error`

**What it means:** The PR itself has structural problems.

**Common causes:**
- PR contains files outside the `manifests/` folder
- PR contains manifests for **more than one package** or **more than one version**
- Modified files that are not part of the manifest submission

**How to fix:** Each PR must contain exactly one package version. Split your submission into separate PRs if needed.

---

#### `Manifest-Installer-Validation-Error`

**What it means:** There are inconsistencies or missing values in the manifest discovered during MSIX package evaluation or installer metadata validation.

**Common causes:**
- `InstallerType` mismatch between what is declared and what is detected

> - Missing or incorrect `PackageFamilyName` for MSIX packages

Reviewer: This can also happen on incorrect MinimumOSVersion for MSIX.

**What it means:** A URL in the manifest failed validation. This can be triggered by Microsoft Defender SmartScreen reputation checks or HTTP error responses (403, 404).

**Common causes:**
- Installer URL returns 404 (file not found) or 403 (forbidden)
- URL has a poor SmartScreen reputation
- URL points to a site flagged for malicious content

**How to fix:**
1. Check that all URLs in the manifest are valid and accessible.
2. If the issue is reputation-based, [submit the URL for review](https://www.microsoft.com/wdsi/filesubmission/).
3. Look at the PR check details to identify which specific URL failed.

---

#### `Validation-HTTP-Error`

**What it means:** The `InstallerUrl` does not use HTTPS.

**How to fix:** Update the `InstallerUrl` to use `https://` instead of `http://`.

---

#### `Validation-Domain`

**What it means:** The domain of the `InstallerUrl` does not match the expected domain for this publisher. WinGet policy requires that installers come directly from the publisher's release location.

**Common causes:**
- Using a third-party CDN, mirror, or download aggregator instead of the official publisher URL
- Using a URL shortener or redirector

> **How to fix:** Use the official download URL from the publisher's website. If the URL is legitimate, add a comment to your PR for investigation.

Reviewer: Might be good to add a best practice note or a tip that adding the `PackageUrl` and ensuring the `InstallerUrl` can be found just by navigating around the website from the `PackageUrl` will help speed up reviews.  

Probably a bit difficult to word, but makes Stephen's job a lot easier if that is the case

---

#### `Validation-Unapproved-URL`

**What it means:** Similar to `Validation-Domain` — the `InstallerUrl` domain is not an approved source for this publisher.

**How to fix:** Use the direct URL from the publisher's official release location.

---

#### `Validation-Indirect-URL`

> **What it means:** The installer URL uses a redirector rather than pointing directly to the publisher's server.

Reviewer: ```suggestion
**What it means:** The installer URL uses a redirect rather than pointing directly to the publisher's server.
```

> **How to fix:** Replace the redirected URL with the direct/final URL from the publisher's server.

Reviewer: Should we provide guidance for users on how to find the final URL? Either using a website like RedirectDetective, or a PowerShell snippet?

<details><summary>PowerShell Snippet</summary>

```powershell
Function Get-UrlResponse {
    Param
    (
        [Parameter(Mandatory = $true, Position = 0)]
        [string] $URL
    )
    try {
        $HTTP_Request = [System.Net.WebRequest]::Create($URL)
        $HTTP_Request.UserAgent = 'Microsoft-Delivery-Optimization/10.1'
        $HTTP_Response = $HTTP_Request.GetResponse()
        $ResponseUri = $HTTP_Response.ResponseUri
        $AbsoluteUrl = $HTTP_Response.ResponseUri.AbsoluteUri
        $HTTP_Status = [int]$HTTP_Response.StatusCode
        $ResponseLength = $HTTP_Response.ContentLength
        $Headers = @{}; $HTTP_Response.Headers.ForEach({ $Headers[$_] = $Http_Response.Headers[$_] })
    } catch {
        $HTTP_Status = 404
    }
    If ($null -eq $HTTP_Response) { $HTTP_Status = 404 }
    Else { $HTTP_Response.Close() }
    return @{
        Url           = $URL
        ResponseUrl   = $AbsoluteUrl
        ResponseCode  = $HTTP_Status
        ContentLength = $ResponseLength
        Headers       = $Headers
        Response      = $ResponseUri
    }
}
```

</details>

---

### Installation Testing Errors

#### `Validation-Unattended-Failed`

**What it means:** The installer did not complete silently — it either timed out or required user interaction.

**Common causes:**

> - Missing or incorrect silent install switches (`/S`, `/silent`, `/quiet`, etc.)

Reviewer: Another common cause - `exe` type specified for `nullsoft` or `burn` or `inno` OR `msi` specified for `wix`. Causes the same behavior as missing or incorrect switches, since WinGet usually handles the switches for those types. Maybe don't need to add as a common cause, but could be the first troubleshooting step to `1. Check your installer type`

| **Internal-Error-URL** | Error during URL reputation validation. |

---

## Moderator-Applied Labels

These labels are applied by [moderators](https://github.com/microsoft/winget-pkgs/blob/master/doc/Moderation.md) during manual review and indicate issues that need author attention.

| Label | Meaning |
|---|---|
| **Error-Hash-Mismatch** | Installer hash doesn't match — same as the automated label. |
| **Error-Installer-Availability** | Installer URL cannot be reached. |
| **Interactive-Only-Installer** | The installer requires user interaction and cannot run silently. This is a **blocking issue**. |
| **License-Blocks-Install** | A license agreement blocks silent installation. This is a **blocking issue**. |
| **Manifest-Content-Incomplete** | Manifest metadata is missing or insufficient. |
| **Manifest-Singleton-Deprecated** | Singleton manifests are no longer accepted — convert to multi-file format. |
| **Version-Parameter-Mismatch** | The version in the manifest doesn't match what the installer writes to the registry. |
| **Portable-Archive** | The installer is a portable archive that may need special handling. This is a **blocking issue**. |

> | **Zip-Binary** | The installer is a zip containing a binary. This is a **blocking issue**. |

Author: This one as well :)

| **Scripted-Application** | The package uses scripts (`.bat`, `.ps1`) as installers, which are **not allowed**. |

> | **Windows-Features** | The installer depends on optional Windows features. This is a **blocking issue**. |

Author: You are correct.

| `Validation-Defender-Error` | Install | ⚠️ Maybe | Fix or submit for false positive review |
| `Validation-Merge-Conflict` | PR | ✅ Yes | Resolve merge conflict |
| `Validation-MSIX-Dependency` | Dependency | ✅ Yes | Add missing framework dependency |
| `Validation-VCRuntime-Dependency` | Dependency | ✅ Yes | Add VC++ runtime dependency |
| `Needs-CLA` | PR | ✅ Yes | Sign the Contributor License Agreement |
| `Internal-Error-*` | Internal | ❌ No | WinGet team will investigate |
| `Policy-Test-*` | Content | ⚠️ Maybe | Additional manual review required |

---

## Getting Help

- **Re-run validation:** A moderator can comment `@wingetbot run` on your PR to re-trigger the validation pipeline.
- **Ask a moderator:** Check [recently closed PRs](https://github.com/microsoft/winget-pkgs/pulls?q=is%3Apr+is%3Aclosed+label%3AModerator-Approved) to find an active moderator and `@mention` them.
- **File an issue:** If you believe the failure is incorrect, [open an issue](https://github.com/microsoft/winget-pkgs/issues/new).
- **Matrix chat:** Join the [WinGet-pkgs Matrix room](https://gitter.im/Microsoft/winget-pkgs) for quick questions.

---

> *For more information, see the [official validation documentation](https://docs.microsoft.com/windows/package-manager/package/winget-validation) and the [Troubleshooting guide](https://github.com/microsoft/winget-pkgs/blob/master/doc/Troubleshoot.md).*

Reviewer: The content in `Troublshoot.md` seems largely duplicated here, I'd suggest removing the troubleshoot doc entirely

| Label | Meaning |
|---|---|

> | **Azure-Pipeline-Passed** | Your manifest passed automated testing and is awaiting moderator approval. |

Reviewer 2: ```suggestion
| **Azure-Pipeline-Passed** | Your manifest passed automated testing. There may be additional manual verification required. |
```

Making the difference between `Azure-Pipeline-Passed` & `Validation-Completed` more apparent. There are a number of PRs that have the `Azure-Pipeline-Passed` but they require manual intervention either by a maintainer or the author https://github.com/microsoft/winget-pkgs/pulls?q=is%3Apr+is%3Aopen+label%3AAzure-Pipeline-Passed+-label%3AValidation-Completed

Attaching a screenshot if there are no PRs matching the filter by the time you're seeing this comment

<img width="1578" height="846" alt="image" src="https://github.com/user-attachments/assets/afb39bd9-a2f6-4a12-bb01-1bc5dbb31de4" />
